File: windows/CPAMPMonitor/MonitorModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;
using System.Text.Json.Nodes;
using System.Windows.Threading;
using CPAMP.MonitorCore;

namespace CPAMP.Windows;

internal sealed class AccountState(Account account)
{
    public Account Account { get; set; } = account;
    public List<QuotaWindow> Windows { get; set; } = [];
    public string Plan { get; set; } = account.Plan;
    public string? Error { get; set; }
    public DateTimeOffset? RetryAfter { get; set; }
    public JsonObject? History { get; set; }
    public bool Enabled { get; set; }
}

internal sealed class MonitorModel : IDisposable
{
    public Configuration Config { get; private set; } = Storage.Load<Configuration>("settings");
    public List<AccountState> Accounts { get; private set; } = [];
    public bool Refreshing { get; private set; }
    public bool Paused { get; private set; }
    public string? Error { get; private set; }
    public string? HistoryError { get; private set; }
    public string? StorageError { get; private set; }
    public DateTimeOffset? LastRefresh { get; private set; }
    public DateTimeOffset? NextRefresh { get; private set; }
    public event Action? Changed;
    public Action<string, string, bool>? Notify;
    private AlertLedger ledger = Storage.Load<AlertLedger>("alerts");
    private readonly DispatcherTimer timer = new() { Interval = TimeSpan.FromSeconds(15) };
    private CancellationTokenSource? request;
    private int generation;
    public double? Lowest => LowestFor(null);
    public double? DisplayLowest => LowestFor(Config.DisplayAccountId);
    public string DisplayLabel => Config.DisplayAccountId is null ? "All accounts minimum" :
        Accounts.FirstOrDefault(a => a.Account.Id == Config.DisplayAccountId)?.Account.Title ?? "Selected account";
    private double? LowestFor(string? accountId)
    {
        if (Paused || Error is not null) return null;
        var rows = Accounts.Where(a => a.Enabled && a.Error is null && (accountId is null || a.Account.Id == accountId));
        return QuotaWindow.MinimumFreshRemaining(rows.SelectMany(a => a.Windows), DateTimeOffset.UtcNow, Config.MaxAge);
    }
    public string Status => Paused ? "Paused" : Refreshing ? "Refreshing" : Config.BaseUrl.Length == 0 ? "Not connected" :
        Error is not null ? "Connection failed" : !Accounts.Any(a => a.Enabled) ? "No monitored accounts" :
        Accounts.Any(a => a.Enabled && (a.Error is not null || a.Windows.Count == 0 ||
            a.Windows.Any(w => !w.Fresh(DateTimeOffset.UtcNow, Config.MaxAge)))) ? "Some data unavailable" :
        Lowest <= Config.Critical ? "Quota critical" : Lowest <= Config.Warning ? "Quota low" : "Monitoring";
    public void Start()
    {
        StorageError = Storage.Warning;
        if (Config.BaseUrl.Length > 0)
            try { Config = Config.Validate(); } catch (MonitorException e) { Error = e.Message; Config = Config with { BaseUrl = "" }; }
        timer.Tick += (_, _) => { Changed?.Invoke(); if (!Paused && NextRefresh <= DateTimeOffset.UtcNow) _ = Refresh(); };
        timer.Start();
        _ = Refresh();
    }
    public void TogglePause()
    {
        Paused = !Paused;
        if (Paused) Cancel(); else _ = Refresh();
        Changed?.Invoke();
    }
    private void Cancel()
    {
        generation++;
        request?.Cancel();
        request = null;
        Refreshing = false;
    }
    public async Task Refresh()
    {
        if (Refreshing || Paused || Config.BaseUrl.Length == 0) return;
        Refreshing = true;
        var current = ++generation;
        using var cts = new CancellationTokenSource();
        request = cts;
        var token = cts.Token;
        Changed?.Invoke();
        try
        {
            var secret = Storage.ReadSecret(Config.BaseUrl);
            if (secret.Length == 0) throw new MonitorException("No management key saved for this address. Open Settings.");
            using var client = new ApiClient(Config.BaseUrl, secret);
            var remote = await client.Accounts(token);
            token.ThrowIfCancellationRequested();
            var old = Accounts.ToDictionary(a => a.Account.Id);
            Accounts = remote.OrderBy(a => { var i = Config.AccountOrder.IndexOf(a.Id); return i < 0 ? int.MaxValue : i; })
                .Select(a => {
                    var row = old.GetValueOrDefault(a.Id) ?? new(a);
                    row.Account = a;
                    row.Enabled = !a.Disabled && !Config.Excluded.Contains(a.Id);
                    return row;
                }).ToList();
            Error = null;
            foreach (var row in Accounts.Where(a => a.Enabled))
            {
                if (row.RetryAfter > DateTimeOffset.UtcNow) continue;
                try
                {
                    var quota = await client.Quota(row.Account, token);
                    token.ThrowIfCancellationRequested();
                    row.Windows = quota.Windows;
                    row.Plan = quota.Plan.Length == 0 ? row.Account.Plan : quota.Plan;
                    row.Error = null;
                    row.RetryAfter = null;
                    if (Config.Notifications)
                        foreach (var window in row.Windows)
                        {
                            var threshold = ledger.Evaluate(Config.BaseUrl + "|" + row.Account.Id + "|" + window.Id,
                                window, Config.Warning, Config.Critical, DateTimeOffset.UtcNow, Config.MaxAge);
                            if (threshold.HasValue) Notify?.Invoke(threshold == Config.Critical ? "Quota critical" : "Quota warning",
                                $"{row.Account.Title}: {window.Title} {window.Remaining:0}% remaining", threshold == Config.Critical);
                            // Preserve deduplication even if a later request is cancelled or the app exits.
                            if (threshold.HasValue) SaveLedger();
                        }
                }
                catch (OperationCanceledException) when (token.IsCancellationRequested) { throw; }
                catch (Exception e)
                {
                    row.Error = ApiClient.SafeMessage(e);
                    if (e is MonitorException { Status: 429 }) row.RetryAfter = DateTimeOffset.UtcNow.AddMinutes(15);
                }
                Changed?.Invoke();
            }
            try
            {
                var history = await client.History(remote, token);
                token.ThrowIfCancellationRequested();
                foreach (var row in Accounts) row.History = history.GetValueOrDefault(row.Account.Id);
                HistoryError = null;
            }
            catch (OperationCanceledException) when (token.IsCancellationRequested) { throw; }
            catch { HistoryError = "Usage history is unavailable."; foreach (var row in Accounts) row.History = null; }
            SaveLedger();
            LastRefresh = DateTimeOffset.Now;
        }
        catch (OperationCanceledException) when (token.IsCancellationRequested) { }
        catch (Exception e) { if (current == generation) Error = ApiClient.SafeMessage(e); }
        finally
        {
            if (current == generation)
            {
                request = null;
                Refreshing = false;
                NextRefresh = DateTimeOffset.Now.AddSeconds(Config.Interval);
                Changed?.Invoke();
            }
        }
    }
    private void SaveLedger()
    {
        try { Storage.Save("alerts", ledger); StorageError = null; }
        catch { StorageError = "Alert history could not be saved. Notifications may repeat after restart."; }
    }
    public void Save(Configuration config, string secret, bool startup)
    {
        config = config.Validate();
        if (secret.Length > 0) Storage.WriteSecret(config.BaseUrl, secret);
        if (Storage.ReadSecret(config.BaseUrl).Length == 0) throw new MonitorException("Enter a management key for this address.");
        Storage.SetStartup(startup);
        Storage.Save("settings", config);
        Cancel();
        if (Config.BaseUrl != config.BaseUrl) { Accounts = []; LastRefresh = null; HistoryError = null; }
        Config = config;
        Error = null;
        Paused = false;
        _ = Refresh();
    }
    public void Dispose() { timer.Stop(); Cancel(); }
}

File: Sources/CPAMPMonitor/MonitorModel.swift
import AppKit
import SwiftUI
import UserNotifications
import ServiceManagement
import MonitorCore

struct AccountState: Identifiable {
    var id: String { account.id }
    var account: Account
    var windows: [QuotaWindow] = []
    var plan = ""
    var error: String?
    var history: [String: Any] = [:]
    var retryAfter: Date?
    var enabled = true
    var fetched: Date?
}

struct DisplayWindowChoice: Identifiable, Hashable {
    let id: String
    let title: String
}

@MainActor final class MonitorModel: ObservableObject {
    @Published var config = Storage.load(Configuration.self, key: "configuration") ?? Configuration()
    @Published var accounts: [AccountState] = []
    @Published var refreshing = false
    @Published var error: String?
    @Published var historyError: String?
    @Published var lastRefresh: Date?
    @Published var nextRefresh: Date?
    @Published var notificationStatus = "未授权"
    @Published var notificationMessage: String?
    @Published var paused = false
    @Published var now = Date()
    var onChange: (() -> Void)?
    var openSettings: (() -> Void)?
    private var timer: Timer?
    private var wakeObserver: NSObjectProtocol?
    private var task: Task<Void, Never>?
    let tunnel = Tunnel()
    private var ledger = Storage.load(AlertLedger.self, key: "alerts") ?? AlertLedger()
    var maxAge: TimeInterval { Double(config.interval * 2 + 60) }
    var configured: Bool { !config.baseURL.isEmpty }
    var freshWindows: [QuotaWindow] {
        guard !paused, error == nil else { return [] }
        return accounts.filter { $0.enabled && $0.error == nil }.flatMap(\.windows).filter { $0.fresh(now: now, maxAge: maxAge) }
    }
    var lowest: Double? { QuotaWindow.minimumFreshRemaining(freshWindows, now: now, maxAge: maxAge) }
    var displayWindowChoices: [DisplayWindowChoice] {
        let rows = accounts.filter { row in
            config.displayAccountID == nil || row.id == config.displayAccountID
        }
        var seen = Set<String>()
        return rows.flatMap { row in
            row.windows.compactMap { window -> DisplayWindowChoice? in
                guard seen.insert(window.id).inserted else { return nil }
                let title = config.displayAccountID == nil
                    ? "\(row.account.providerName) · \(window.title)" : window.title
                return DisplayWindowChoice(id: window.id, title: title)
            }
        }
    }
    var displayLowest: Double? {
        guard !paused, error == nil else { return nil }
        let rows = accounts.filter { row in
            row.enabled && row.error == nil && (config.displayAccountID == nil || row.id == config.displayAccountID)
        }
        return QuotaWindow.minimumFreshRemaining(rows.flatMap(\.windows), id: config.displayWindowID,
                                                 now: now, maxAge: maxAge)
    }
    var displayAccountLabel: String {
        guard let id = config.displayAccountID else { return "所有账号" }
        return accounts.first(where: { $0.id == id })?.account.title ?? "指定账号"
    }
    var displayWindowLabel: String {
        guard config.displayWindowID != nil else { return "所有额度" }
        return displayWindowChoices.first(where: { $0.id == config.displayWindowID })?.title ?? "指定额度"
    }
    var displayLabel: String { "\(displayAccountLabel) · \(displayWindowLabel)" }
    var hasProblems: Bool {
        error != nil || accounts.contains { $0.enabled && ($0.error != nil || $0.windows.isEmpty || $0.windows.contains { !$0.fresh(now: now, maxAge: maxAge) }) }
    }
    var statusText: String {
        if paused { return "监控已暂停" }
        if refreshing { return "正在刷新" }
        if !configured { return "尚未连接" }
        if error != nil { return "连接失败" }
        if hasProblems { return "部分数据不可用" }
        if let lowest, lowest <= Double(config.critical) { return "额度紧张" }
        if let lowest, lowest <= Double(config.warning) { return "额度偏低" }
        if !accounts.contains(where: \.enabled) { return "无监控账号" }
        return "监控中"
    }

    func start() {
        timer = Timer.scheduledTimer(withTimeInterval: 15, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.now = Date()
                self.onChange?()
                if !self.paused, let next = self.nextRefresh, next <= self.now { self.refresh() }
            }
        }
        wakeObserver = NSWorkspace.shared.notificationCenter.addObserver(forName: NSWorkspace.didWakeNotification, object: nil, queue: .main) { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self else { return }
                self.now = Date()
                if !self.paused, self.lastRefresh == nil || Date().timeIntervalSince(self.lastRefresh!) > 30 { self.refresh() }
            }
        }
        Task {
            if configured && config.notifications { await requestNotifications() }
            else { await updateNotificationStatus() }
        }
        refresh()
    }

    func refresh() {
        guard configured, !refreshing, !paused else { return }
        refreshing = true
        onChange?()
        task = Task { await performRefresh() }
    }

    private func orderedAccounts(_ remote: [Account]) -> [Account] {
        guard !config.accountOrder.isEmpty else { return remote }
        let byID = Dictionary(uniqueKeysWithValues: remote.map { ($0.id, $0) })
        let preferred = config.accountOrder.compactMap { byID[$0] }
        let known = Set(preferred.map(\.id))
        return preferred + remote.filter { !known.contains($0.id) }
    }

    private func performRefresh() async {
        defer {
            refreshing = false
            now = Date()
            nextRefresh = now.addingTimeInterval(Double(config.interval))
            onChange?()
        }
        do {
            try await tunnel.prepare(config)
            let secret = try Storage.credential(config.credentialID)
            guard !secret.isEmpty else { throw MonitorError("尚未保存管理密钥，请打开设置。") }
            let client = try APIClient(baseURL: config.baseURL, key: secret)
            let remote = try await client.accounts()
            let old = Dictionary(accounts.map { ($0.id, $0) }, uniquingKeysWith: { first, _ in first })
            accounts = orderedAccounts(remote).map { account in
                var row = old[account.id] ?? AccountState(account: account)
                row.account = account
                row.enabled = !account.disabled && !config.excluded.contains(account.id)
                return row
            }
            error = nil
            for index in accounts.indices where accounts[index].enabled {
                if let retry = accounts[index].retryAfter, retry > Date() { continue }
                do {
                    let quota = try await client.quota(accounts[index].account)
                    accounts[index].windows = quota.windows
                    accounts[index].plan = quota.plan
                    accounts[index].error = nil
                    accounts[index].retryAfter = nil
                    accounts[index].fetched = Date()
                    await notifyIfNeeded(accounts[index])
                } catch {
                    accounts[index].error = safeMessage(error)
                    if (error as? MonitorError)?.status == 429 {
                        accounts[index].retryAfter = Date().addingTimeInterval(900)
                    }
                }
                now = Date()
                onChange?()
            }
            do {
                let history = try await client.history(remote)
                for index in accounts.indices { accounts[index].history = history[accounts[index].id] ?? [:] }
                historyError = nil
            } catch { historyError = "历史用量暂不可用" }
            lastRefresh = Date()
        } catch { self.error = safeMessage(error) }
    }

    func safeMessage(_ error: Error) -> String {
        if let e = error as? MonitorError { return e.message }
        if let e = error as? URLError {
            return e.code == .timedOut ? "网络请求超时。" : "网络连接失败（\(e.code.rawValue)）。"
        }
        return "响应解析失败，请检查服务版本与地址。"
    }

    private func notifyIfNeeded(_ row: AccountState) async {
        guard config.notifications, !paused, row.enabled else { return }
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        guard settings.authorizationStatus == .authorized || settings.authorizationStatus == .provisional else { return }
        for window in row.windows {
            var candidate = ledger
            let key = config.credentialID + "|" + row.id + "|" + window.id
            let threshold = candidate.evaluate(key: key, window: window, warning: config.warning,
                                               critical: config.critical, now: Date(), maxAge: maxAge)
            if let threshold {
                let content = UNMutableNotificationContent()
                content.title = threshold == config.critical ? "CPAMP 额度紧急提醒" : "CPAMP 额度提醒"
                content.body = "\(row.account.title) · \(window.title) 剩余 \(Int(window.remaining ?? 0))%"
                content.sound = .default
                do {
                    try await UNUserNotificationCenter.current().add(.init(identifier: UUID().uuidString, content: content, trigger: nil))
                } catch { notificationMessage = "通知发送失败，请检查系统通知设置。"; continue }
            }
            ledger = candidate
            Storage.save(ledger, key: "alerts")
        }
    }

    func updateNotificationStatus() async {
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        switch settings.authorizationStatus {
        case .authorized, .provisional: notificationStatus = "已授权"
        case .denied: notificationStatus = "已被系统关闭"
        default: notificationStatus = "未授权"
        }
    }

    func requestNotifications(test: Bool = false) async {
        do {
            let allowed = try await UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge])
            await updateNotificationStatus()
            guard allowed else { notificationMessage = "请在系统设置 > 通知中允许 CPAMP Monitor。"; return }
            if test {
                let content = UNMutableNotificationContent()
                content.title = "CPAMP Monitor 测试通知"
                content.body = "额度通知已连接。这是一条测试，不代表额度不足。"
                content.sound = .default
                try await UNUserNotificationCenter.current().add(.init(identifier: UUID().uuidString, content: content, trigger: nil))
                notificationMessage = "测试通知已交给系统；显示方式受专注模式和通知设置影响。"
            }
        } catch { notificationMessage = "无法发送通知，请检查系统权限。" }
    }

    func save(_ newConfig: Configuration, secret: String, login: Bool) async throws {
        guard !refreshing else { throw MonitorError("请等待当前刷新完成。") }
        let validated = try newConfig.validated()
        if !secret.isEmpty { try Storage.storeCredential(secret, server: validated.credentialID) }
        guard try !Storage.credential(validated.credentialID).isEmpty else { throw MonitorError("该地址尚未保存管理密钥。") }
        let changedServer = validated.credentialID != config.credentialID
        if login && SMAppService.mainApp.status != .enabled { try SMAppService.mainApp.register() }
        if !login && SMAppService.mainApp.status == .enabled { try await SMAppService.mainApp.unregister() }
        config = validated
        Storage.save(config, key: "configuration")
        if changedServer { accounts = []; lastRefresh = nil; error = nil }
        if config.notifications { await requestNotifications() }
        paused = false
        refresh()
    }

    func togglePause() {
        paused.toggle()
        if !paused { refresh() }
        onChange?()
    }

    func openDashboard() {
        guard let url = try? Configuration.normalizeURL(config.baseURL) else { return }
        NSWorkspace.shared.open(url.appendingPathComponent("management.html"))
    }
}
